### PR TITLE
Reworked SMS Response Code generate to prevent infinite loops.

### PR DESCRIPTION
**Pull request by cabal95**
## Contributor Agreement
_Have you filled out and sent your [Spark Contributor Agreement](http://www.rockrms.com/Content/RockExternal/Misc/Contributor%20Agreement.pdf) to secretary [at] sparkdevnetwork.org?_

**Yes**

## Context
_What is the problem you encountered that lead to you creating this pull request?_

Issue #2552, when all 899 available response codes had been used in the past 30 days the system would loop forever trying to generate a new code.

## Goal
_What will this pull request achieve and how will this fix the problem?_

Increase number of available codes, decrease window for a code being considered "used" and rework the loop to have a final escape.

## Strategy
_How have you implemented your solution?_

System now supports up to 5-digit response codes. It generates them in a "bucket system" of 100 at a time, between 100 and 90,000 (skipping `666` and `911`). These chunks are generated using the "smallest number available" logic. Once the chunk list is generated it is randomized. Because of this "bucket system" the randomized list will be at most 100 sequential numbers, so we may want to increase the bucket size if we find there is not enough entropy.

This randomized list is then stored in the cache and re-used until the list has been depleted, at which time a new list is generated.

When trying to find the next single available code, it will try 1,000 times to find a code before giving up. An exception is now thrown if that 1,000 limit is reached or no codes are found to be available when generating the cached list.

## Performance

For testing performance, I pre-generated codes 100 - 3,927 as recently used response codes to determine how much time would be used by the system to process incoming SMS messages.

* Took 5,673ms to find recently used codes (3,827 database rows).
* Took 51ms to find first available chunk of codes (i.e. 3,928 - 3,999).
* Took 0ms to randomize list.

These timings only applied on the first (simulated) incoming SMS message. Subsequent messages took <2ms to pull from the cache and check the database to ensure that one single code was still available.

## Tests
_If your code is a new method or function (that doesn't need a mock database or SqlServerTypes library) and can be Xunit tested [see example](https://github.com/SparkDevNetwork/Rock/blob/develop/Rock.Tests/Rock/Lava/RockFiltersTests.cs) be sure your pull request includes the corresponding unit tests in the Rock.Tests project. In all cases *you* MUST test your code before submitting a pull request._

n/a

## Possible Implications
_What could this change potentially impact? Are there any security considerations? Where could this potentially affect backwards compatibility?_

n/a

## Screenshots
_Provide us some screenshots if your pull request either alters existing UI or provides new UI. Arrows and labels are helpful._

n/a

## Documentation
_If your change effects the UI or needs to be documented in one of the existing [user guides](http://www.rockrms.com/Learn/Documentation), please provide the brief write-up here:_

n/a

## Migrations
Should your pull request require a migration, please exclude the migration from the Rock.Migration project, but submit it in your pull request. Please add a note to your pull request that provides a heads up that a migration file is present.

n/a

### CODE REVIEW

**Comment by azturner** [comment]
@cabal95 In testing/looking at this, it does not appear that the codes are being randomized at all. First two codes I got were @100 and @101. I'm assuming the `availableCodes.OrderBy( c => Guid.NewGuid() ).ToList();` is what you're using to try and randomize them, but that does not seem to do so.

**Comment by cabal95** [comment]
@azturner Hmm. It did work for me. It's possible it was an incredible fluke.  I did test that style of code again just now on https://dotnetfiddle.net and it worked. That was also the recommended method I found on a few stack overflow posts, but I suppose it's possible something with a slightly different .NET version or something. Is it easy enough for you to test again with a debugger breakpoint there to see what is actually in the list after randomization? I'll see if I can test again on my system, but I'm not sure that branch will run against my database anymore.

Code used to test on dotnetfiddle.net:
```
using System;
using System.Collections.Generic;
using System.Linq;
					
public class Program
{
	public static void Main()
	{
		var list = new List<int> { 1, 2, 3, 4, 5 };
		
		Console.WriteLine( string.Join(", ", list.OrderBy( c => Guid.NewGuid() ) ) );
	}
}
```

File: Rock/Communication/Medium/Sms.cs [hunk 1]
```diff
@@ -179,38 +197,92 @@ private void CreateCommunication( int fromPersonAliasId, string fromPersonName,
...
+                    .Where( i => !blacklist.Contains( i ) )
+                    .Select( i => string.Format( "@{0}", i ) )
+                    .Where( c => !activeCodes.Contains( c ) )
+                    .ToList();
+
+                if ( availableCodes.Any() )
+                {
+                    availableCodes.OrderBy( c => Guid.NewGuid() ).ToList();
+                    return availableCodes;
+                }
```
Comment (azturner) [inline Rock/Communication/Medium/Sms.cs]: @cabal95 The orderby actually works, but the reordered list is not saved to a new list. If we just change these two lines to...
`return availableCodes.OrderBy( c => Guid.NewGuid() ).ToList();`
it works fine. I'll just pull this in and make that change.

Comment (cabal95) [reply Rock/Communication/Medium/Sms.cs]: Oh, that will do it. I must have had that in two lines when I tested and then refactored it into two right before committing without retesting. Sorry 'bout that!

Comment (azturner) [reply Rock/Communication/Medium/Sms.cs]: @cabal95 No Prob!  Thanks for doing all these pull requests. It's awesome!

**Pull request merged**